src/Infrastructure/Common/Middlewares/SecurityMiddleware.cs

```diff
@@ -0,0 +1,97 @@
+﻿using DN.WebApi.Application.Identity.Exceptions;
+using DN.WebApi.Application.Identity.Interfaces;
+using DN.WebApi.Application.Settings;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Options;
+using System.Collections.Concurrent;
+using System.Net;
+
+namespace DN.WebApi.Infrastructure.Middlewares;
+
+public class SecurityMiddleware : IMiddleware
+{
+    private static readonly ConcurrentDictionary<string, RequestFailed> _blackListIp = new();
+
+    private readonly ICurrentUser _currentUser;
+    private readonly MiddlewareSettings _settings;
+    private DateTime _nextClear;
+
+    public SecurityMiddleware(IOptions<MiddlewareSettings> midsettings, ICurrentUser currentUser)
+    {
+        _settings = midsettings.Value;
+        _currentUser = currentUser;
+        _nextClear = DateTime.UtcNow.AddMinutes(_settings.BlackListMinutes);
+    }
+
+    public async Task InvokeAsync(HttpContext context, RequestDelegate next)
+    {
+        if (_nextClear <= DateTime.UtcNow)
+        {
+            _nextClear = DateTime.UtcNow.AddHours(1);
+            foreach (var target in _blackListIp.ToArray())
+            {
+                if (target.Value.NextReset <= DateTime.UtcNow)
+                    _blackListIp.TryRemove(target.Key, out _);
+            }
+        }
+
+        // avoid to block auth users
+        if (!_currentUser.IsAuthenticated())
+        {
+            // avoid x-forwarded tag, cracker can put fake input to avoid block
+            string currentIp = context.Connection.RemoteIpAddress.MapToIPv4().ToString();
+            if (_blackListIp.TryGetValue(currentIp, out var requestFailed))
```

Comment:
isn't this gonna be a problem when running behind a reverse proxy or load balancer?